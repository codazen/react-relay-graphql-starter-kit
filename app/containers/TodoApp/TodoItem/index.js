/* @flow */

import React from 'react';
import Relay from 'react-relay';
import ListItem from '../../../components/ListItem';

type Props = {
  todo: {
    content: string,
    id: string,
  },
};

class TodoItem extends React.Component {
  constructor(props: Props) { // eslint-disable-line
    super(props);
  }

  render() {
    return (
      <ListItem
        content={this.props.todo.content}
        id={this.props.todo.id}
      />
    );
  }
}

export default Relay.createContainer(TodoItem, {
  fragments: {
    todo: () => Relay.QL`
      fragment on Todo {
        id
        content
      }
    `,
  },
});
