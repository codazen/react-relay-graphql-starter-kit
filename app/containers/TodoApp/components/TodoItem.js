/* @flow */

import React from 'react';
import Relay from 'react-relay';
import ListItem from 'components/ListItem';
import RemoveTodoMutation from '../mutations/RemoveTodoMutation';
import UpdateTodoMutation from '../mutations/UpdateTodoMutation';

type Todo = {
  id: string,
  content: string,
};

type Props = {
  todo: Todo,
  user: {
    id: string,
  },
  relay: {
    commitUpdate: Function, // function to make mutation call to relay store
  },
};

/**
 * ListItem wrapper for TodoList implementation
 * Parent component: TodoList
 * Child component: ListItem
 */
class TodoItem extends React.Component {

  props: Props;

  /**
   * removeTodo function passed to the ListItem component
   * @param {id: string} id of todo to be removed
   */
  handleRemoveTodo = (id: string) => {
    this.props.relay.commitUpdate(
      new RemoveTodoMutation({ id, user: this.props.user }),
    );
  }

  /**
   * updateTodo function passed to the ListItem component
   * @param {todo: Todo} todo node containing id and content
   */
  handleUpdateTodo = (todo: Todo) => {
    this.props.relay.commitUpdate(
      new UpdateTodoMutation({ todo: this.props.todo, content: todo.content }),
    );
  }

  render() {
    const {
      content,
      id,
    } = this.props.todo;
    return (
      <ListItem
        content={content}
        id={id}
        removeItem={this.handleRemoveTodo}
        updateItem={this.handleUpdateTodo}
      />
    );
  }
}

exports.TodoItem = TodoItem;

export default Relay.createContainer(TodoItem, {
  fragments: {
    todo: () => Relay.QL`
      fragment on Todo {
        id
        content
        ${UpdateTodoMutation.getFragment('todo')}
      }
    `,
    user: () => Relay.QL`
      fragment on User {
        id
        ${RemoveTodoMutation.getFragment('user')}
      }
    `,
  },
});
