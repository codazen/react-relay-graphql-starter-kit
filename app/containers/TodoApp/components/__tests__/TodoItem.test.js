import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { makeExecutableSchema, addMockFunctionsToSchema, MockList } from 'graphql-tools';
import { graphql } from 'graphql';
import casual from 'casual';
import schemaString from 'data/schema.graphql';
import { TodoItem } from '../TodoItem';
import RemoveTodoMutation from '../../mutations/RemoveTodoMutation';
import UpdateTodoMutation from '../../mutations/UpdateTodoMutation';

// setup React as global variable
global.React = React;

// use consistent seed for snapshot matching
casual.seed(123);

// create schema based on server/data/schema.graphql
const schema = makeExecutableSchema({ typeDefs: schemaString });

// mock data
const numListItems = 10;
const mocks = {
  TodoConnection: () => ({
    edges: () => new MockList(numListItems),
  }),
  Todo: () => ({
    id: () => casual.integer(0, 1000).toString(),
    content: () => casual.string,
  }),
};

// add mocks to schema in place
addMockFunctionsToSchema({
  schema,
  mocks,
});

describe('TodoItem component', () => {
  let todoItem;
  let todo;
  let user;
  const query = `query
    TodoItem {
      user {
        id
        todos(first: 10) {
          edges {
            node {
              id
              content
            }
          }
        }
      }
    }
  `;

  // mock mutation calls and use _unresolvedProps to detect failure
  const RemoveTodoStatus = jest.fn(mutation => mutation._unresolvedProps.id); // eslint-disable-line
  const UpdateTodoStatus = jest.fn(mutation => mutation._unresolvedProps.content); // eslint-disable-line
  const mockSuccessFn = jest.fn((onSuccess, transaction) => onSuccess(transaction));
  const mockFailureFn = jest.fn((onFailure, transaction) => onFailure(transaction));

  const relay = {
    commitUpdate: jest.fn((mutation, callbacks) => {
      if (mutation instanceof RemoveTodoMutation) {
        return RemoveTodoStatus(mutation)
          ? mockSuccessFn(callbacks ? callbacks.onSuccess : () => {}, {})
          : mockFailureFn(callbacks ? callbacks.onFailure : () => {}, {});
      } else if (mutation instanceof UpdateTodoMutation) {
        return UpdateTodoStatus(mutation)
          ? mockSuccessFn(callbacks ? callbacks.onSuccess : () => {}, {})
          : mockFailureFn(callbacks ? callbacks.onFailure : () => {}, {});
      }
      return {};
    }),
  };

  beforeEach(async () => {
    await graphql(schema, query).then((result) => {
      todo = result.data.user.todos.edges[0].node;
      user = result.data.user;
      todoItem = shallow(<TodoItem todo={todo} user={user} relay={relay} />);
      RemoveTodoStatus.mockClear();
      UpdateTodoStatus.mockClear();
      mockSuccessFn.mockClear();
      mockFailureFn.mockClear();
    });
  });

  it('Renders correctly', () => {
    expect(shallowToJson(todoItem)).toMatchSnapshot();
  });

  describe('Unit test: handleRemoveTodo', () => {
    it('Success', () => {
      const id = todoItem.instance().props.todo.id;
      todoItem.instance().handleRemoveTodo(id);
      expect(todoItem.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(RemoveTodoStatus).toHaveBeenCalled();
      expect(mockSuccessFn).toHaveBeenCalled();
    });

    it('Fails', () => {
      const id = undefined;
      todoItem.instance().handleRemoveTodo(id);
      expect(todoItem.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(RemoveTodoStatus).toHaveBeenCalled();
      expect(mockFailureFn).toHaveBeenCalled();
    });
  });

  describe('Unit test: handleUpdateTodo', () => {
    it('Success', () => {
      const todoUpdate = todoItem.instance().props.todo;
      todoItem.instance().handleUpdateTodo(todoUpdate);
      expect(todoItem.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(UpdateTodoStatus).toHaveBeenCalled();
      expect(mockSuccessFn).toHaveBeenCalled();
    });

    it('Fails', () => {
      const todoUpdate = todoItem.instance().props.todo;
      todoUpdate.content = undefined;
      todoItem.instance().handleUpdateTodo(todoUpdate);
      expect(todoItem.instance().props.relay.commitUpdate).toHaveBeenCalled();
      expect(UpdateTodoStatus).toHaveBeenCalled();
      expect(mockFailureFn).toHaveBeenCalled();
    });
  });
});
