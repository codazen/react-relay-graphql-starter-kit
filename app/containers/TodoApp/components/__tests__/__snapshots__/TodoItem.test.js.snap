exports[`TodoItem component Renders correctly 1`] = `
<ListItem
  content="quo quis molestiae tempora est omnis porro"
  id="696"
  removeItem={[Function]}
  updateItem={[Function]} />
`;
